wrappers/rust/icicle-core/src/curve.rs

```diff
@@ -191,3 +196,161 @@ where
         }
     }
 }
+
+#[macro_export]
+macro_rules! impl_curve {
+    (
+        $scalar_limbs:literal,
+        $base_limbs:literal,
+    ) => {
+        #[derive(Debug, PartialEq, Copy, Clone)]
+        pub struct ScalarCfg {}
+
+        impl FieldConfig for ScalarCfg {
+            #[cfg(feature = "arkworks")]
+            type ArkField = Fr;
+        }
+
+        const SCALAR_LIMBS: usize = $scalar_limbs;
+
+        pub type ScalarField = Field<SCALAR_LIMBS, ScalarCfg>;
+
+        extern "C" {
+            fn GenerateScalars(scalars: *mut ScalarField, size: usize);
+        }
+
+        pub(crate) fn generate_random_scalars(size: usize) -> Vec<ScalarField> {
+            let mut res = vec![ScalarField::zero(); size];
+            unsafe { GenerateScalars(&mut res[..] as *mut _ as *mut ScalarField, size) };
+            res
+        }
+
+        #[derive(Debug, PartialEq, Copy, Clone)]
+        pub struct BaseCfg {}
+
+        impl FieldConfig for BaseCfg {
+            #[cfg(feature = "arkworks")]
+            type ArkField = Fq;
+        }
+
+        pub const BASE_LIMBS: usize = $base_limbs;
+
+        #[derive(Debug, PartialEq, Copy, Clone)]
+        pub struct CurveCfg {}
+
+        pub type BaseField = Field<BASE_LIMBS, BaseCfg>;
+        pub type G1Affine = Affine<BaseField, CurveCfg>;
+        pub type G1Projective = Projective<BaseField, CurveCfg>;
+
+        extern "C" {
+            fn Eq(point1: *const c_void, point2: *const c_void) -> c_uint;
+            fn ToAffine(point: *const c_void, point_out: *mut c_void);
+            fn GenerateProjectivePoints(points: *mut G1Projective, size: usize);
+            fn GenerateAffinePoints(points: *mut G1Affine, size: usize);
+        }
+
+        impl CurveConfig for CurveCfg {
+            type BaseField = BaseField;
+            type ScalarField = ScalarField;
+            type Affine = G1Affine;
+            type Projective = G1Projective;
+
+            fn eq_proj(point1: *const c_void, point2: *const c_void) -> c_uint {
+                unsafe { Eq(point1, point2) }
+            }
+
+            fn to_affine(point: *const c_void, point_out: *mut c_void) {
+                unsafe { ToAffine(point, point_out) };
+            }
+
+            #[cfg(feature = "arkworks")]
+            type ArkSWConfig = ArkG1Config;
+        }
+
+        pub(crate) fn generate_random_projective_points(size: usize) -> Vec<G1Projective> {
+            let mut res = vec![G1Projective::zero(); size];
+            unsafe { GenerateProjectivePoints(&mut res[..] as *mut _ as *mut G1Projective, size) };
+            res
+        }
+
+        pub(crate) fn generate_random_affine_points(size: usize) -> Vec<G1Affine> {
+            let mut res = vec![G1Affine::zero(); size];
+            unsafe { GenerateAffinePoints(&mut res[..] as *mut _ as *mut G1Affine, size) };
+            res
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! impl_curve_tests {
```

Comment:
A general note about all tests: earlier we discussed that it'd be nice to have all tests implemented in generic fashion outside of macros. I see at least one issue with this in functions `generate_random_scalars` and `generate_random_*_points` which are themselves implemented inside a macro. Is this the only blocker or is there anything else?

Reply:
Let's leave it as TODO for this PR maybe?